src/mavsdk/core/callback_list.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <utility>
+#include <vector>
+
+namespace mavsdk {
+
+template<typename... Args> class CallbackList;
+
+template<typename... Args> class Handle {
+public:
+    explicit Handle(uint64_t id) : _id(id) {}
```

Comment:
There should perhaps be a way to couple the handle to a callbacklist, so they can't be shuffled(use the wrong handle with the wrong cblist). Something like a random key that the cblist generates and gives to all its handles that is also verified when using the handle to remove it. That would prevent people from trying to remove callback with the wrong handle at runtime. You could also use the function signature and that'd add some more protection at compile time.

Reply:
Anther thing that's perhaps needed is a way to know if the handle is 'alive', of if it has been used to free up the callback already.

Reply:
> Something like a random key that the cblist generates and gives to all its handles that is also verified when using the handle to remove it.

I thought about that but then decided it was probably overkill because the handle is already type checked, as the handle depends on the callback `Args`. The only way to mix it up is if you have e.g. two instances of mavsdk and subscribe to two methods with the same signature. You could - at runtime - avoid it using a random key, indeed.

> Anther thing that's perhaps needed is a way to know if the handle is 'alive',

Hm, you could yes, What would you need it for? I would argue that if you just want to unsubscribe you can just unsubscribe multiple times and it will ignore it. Alternatively, we could make it so that you have to `std::move` the handle into the unsubscribe method, so that the compiler might warn you, if you use it again afterwards, so:

```
void unsubscribe_foo(Handle&& handle);
```

Reply:
> Hm, you could yes, What would you need it for?

I think the obvious way to use it would be to let it carry some state, then you could assign some meaning to if(handle.empty()) do_something(); later in the code. But I can also see why you might explicitly not want to allow such meanings to sit on top of the handle and say use another variable to do it. Its a nice to have.

Reply:
Is there another use-case than doing something like below?

```c++
if (handle.empty()) {
    unsubscribe(handle);
}
```

Because then, I think it's nicer to just say that calling `unsubscribe(handle)` with an already unsubscribed handle is simply a noop.

I can't imagine a scenario where I would do:

```
if (handle.empty()) {
  // do some complex logic here that relies on the handle having been unsubscribed already
}
```

:thinking: